pkg/ccl/importccl/import_pgdump.go

```diff
@@ -0,0 +1,503 @@
+		if err != nil {
+			return err
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, s)
+	case *tree.AlterTable:
+		// If the `ALTER TABLE` statement has an explicit database name, replace it
+		// with the temporary pgdump database being imported into.
+		if stmt.Table.HasExplicitCatalog() {
+			if stmt.Table.Parts[2] != handler.dumpDatabaseName {
+				return errors.AssertionFailedf("catalog name %s does not match dump target database name %s",
+					stmt.Table.Parts[2], handler.dumpDatabaseName)
+			}
+			stmt.Table.Parts[2] = importTempPgdumpDB
+		}
+		for _, cmd := range stmt.Cmds {
+			switch cmd := cmd.(type) {
+			case *tree.AlterTableAddConstraint:
+				switch con := cmd.ConstraintDef.(type) {
+				case *tree.ForeignKeyConstraintTableDef:
+					// TODO(adityamaru): handle FKs and fk skip option.
+					if con.Table.ExplicitCatalog {
+						con.Table.CatalogName = importTempPgdumpDB
+					}
+				default:
+					// TODO(adityamaru): confirm that not other constraint can have a
+					// qualified table name.
+				}
+			case *tree.AlterTableSetDefault:
+			case *tree.AlterTableSetVisible:
+			case *tree.AlterTableAddColumn:
+				if cmd.IfNotExists {
+					return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported statement: %s", stmt))
+				}
+			case *tree.AlterTableSetNotNull:
+			default:
+				return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported statement: %s", stmt))
+			}
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, formatPostgresStatement(stmt))
+	case *tree.CreateIndex:
+		// If the `CREATE INDEX` specifies an explicit database name, replace it
+		// with the temporary pgdump database being imported into.
+		s, err := rewritePostgresStatementTableName(stmt, handler.dumpDatabaseName)
+		if err != nil {
+			return err
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, s)
+	case *tree.AlterSchema:
+		// If the schema specifies an explicit database name, replace it with the
+		// temporary pgdump database being imported into.
+		if stmt.Schema.ExplicitCatalog {
+			if stmt.Schema.CatalogName != tree.Name(handler.dumpDatabaseName) {
+				return errors.AssertionFailedf("catalog name %s does not match dump target database name %s",
+					stmt.Schema.CatalogName, handler.dumpDatabaseName)
+			}
+			stmt.Schema.CatalogName = importTempPgdumpDB
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, formatPostgresStatement(stmt))
+	case *tree.CreateSequence:
+		s, err := rewritePostgresStatementTableName(stmt, handler.dumpDatabaseName)
+		if err != nil {
+			return err
+		}
+		handler.bufferedDDLStmts = append(handler.bufferedDDLStmts, s)
+	case *tree.AlterTableOwner:
+		return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported statement: %s", stmt))
+	case *tree.AlterSequence:
+		return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported %T statement: %s", stmt, stmt))
+	// Some SELECT statements mutate schema. Search for those here.
+	case *tree.Select:
```

Comment:
perhaps worth adding a test case that follows this code branch. I'm not totally clear on what kinds of select statements are cool and which are not.